cmd/lp-colouring/colouring.go

```diff
@@ -0,0 +1,180 @@
+package main
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ScottSallinen/lollipop/graph"
+)
+
+const EmptyColour = math.MaxUint32
+
+type VertexProperty struct {
+	NbrColours sync.Map
+	WaitCount  int64
+	Colour     uint32
+}
+
+func (p *VertexProperty) String() string {
+	s := fmt.Sprintf("{%d,%d,[", p.Colour, p.WaitCount)
+	p.NbrColours.Range(func(key, value any) bool {
+		s += fmt.Sprintf("%d:%d,", key, value)
+		return true
+	})
+	return s + "]}"
+}
+
+func hash(id uint32) (hash uint32) {
+	// TODO: dummy hash function
+	return id
+}
+
+func comparePriority(p1, p2 uint32, id1, id2 uint32) bool {
+	return p1 > p2 || (p1 == p2 && id1 > id2)
+}
+
+// findFirstUnused finds the smallest unused colour, modifies colours, O(n)
+func findFirstUnused(colours []uint32) (firstUnused uint32) {
+	// There must be an available colour <= length
+	// similar to https://leetcode.com/problems/first-missing-positive/
+	length := uint32(len(colours))
+	for i := range colours {
+		for colours[i] < length && colours[i] != colours[colours[i]] {
+			colours[i], colours[colours[i]] = colours[colours[i]], colours[i]
+		}
+	}
+	for i, colour := range colours {
+		if uint32(i) != colour {
+			return uint32(i)
+		}
+	}
+	return length
+}
+
+func MessageAggregator(target, source *graph.Vertex[VertexProperty], data float64) (newInfo bool) {
+	colour := uint32(data)
+	// info(fmt.Sprintf("MessageAggregator target.Id=%v source.Id=%v data=%v colour=%v", target.Id, source.Id, data, colour))
+	target.Property.NbrColours.Store(source.Id, colour)
+
+	if atomic.LoadInt64(&target.Property.WaitCount) > 0 {
+		newWaitCount := atomic.AddInt64(&target.Property.WaitCount, -1)
+		return newWaitCount <= 0 // newWaitCount might go below 0
+	}
+
+	// If we have priority, there is no need to update check our colour
+	if comparePriority(hash(target.Id), hash(source.Id), target.Id, source.Id) {
+		return false
+	}
+
+	return true
+}
+
+func AggregateRetrieve(target *graph.Vertex[VertexProperty]) float64 {
+	return 0
+}
+
+func OnInitVertex(g *graph.Graph[VertexProperty], vidx uint32) {
```

Comment:
Hi Scott,

I tried making it compute the wait count in OnVisitVertex, but because the framework does not guarantee a vertex is visited before its message aggregator runs (the MessageAggregator might see an uninitialized wait count), this task turns out to be non-trivial.

A few possible ways to fix it:
1. Initialize the wait count in MessageAggregator. This is weird as the MessageAggregator is called by the sending vertex's thread.
2. In MessageAggregator, instead of decrementing the WaitCount, increment a separate field called MessageCount. And the OnVisitVertex function will compare these two fields before proceeding.
3. Have a separate field to indicate whether the WaitCount is initialized, allowing it to be decremented before computing its initial value. For example, we can set it to MaxInt64 in OnInitVertex, allows it to decrease in MessageAggregator, and adjust it by (-MaxInt64 + TotalWaitCount ) when initializing it.

Anyways, I think parallelizing the execution of the initializations is much easier than fixing this issue.

Reply:
Hm, good observations. I agree with your conclusion.
I will parallelize the init function instead, so no need to make changes here then.

Reply:
Done in my latest commit to the main branch.
p.s., seems there are some incomplete changes in your latest commit on this branch; perhaps not adjusting other algorithms with the changes to the aggregator.

Reply:
Oops... Sorry. I have fixed it.